filestore/filestore.go

```diff
@@ -129,6 +129,15 @@ func (f *Filestore) Has(c *cid.Cid) (bool, error) {
 }
 
 func (f *Filestore) Put(b blocks.Block) error {
+	has, err := f.Has(b.Cid())
+	if err != nil {
+		return err
+	}
+
+	if has {
+		return nil
+	}
+
 	switch b := b.(type) {
```

Comment:
@whyrusleeping I don't know what you have in mind for the filestore.  But I very specially designed my filestore so that writes go though.  This way if a file is moved it can simply be readded and any invalid blocks are automatically fixed.